src/main/java/htsjdk/tribble/gff/Gff3Codec.java

```diff
@@ -0,0 +1,432 @@
+                break;
+
+            case FASTA_DIRECTIVE:
+                reachedFasta = true;
+                break;
+
+            default:
+                throw new IllegalArgumentException( "directive " + directive + " not handled by parser.");
+
+        }
+    }
+
+    /**
+     * move active top level features to featuresToFlush.  clear active features.
+     */
+    private void prepareToFlushFeatures() {
+        featuresToFlush.addAll(activeTopLevelFeatures);
+        activeFeaturesWithIDs.clear();
+        activeTopLevelFeatures.clear();
+    }
+
+    @Override
+    public LineIterator makeSourceFromStream(final InputStream bufferedInputStream) {
+        return new LineIteratorImpl(new SynchronousLineReader(bufferedInputStream));
+    }
+
+    @Override
+    public LocationAware makeIndexableSourceFromStream(final InputStream bufferedInputStream) {
+        return new AsciiLineReaderIterator(AsciiLineReader.from(bufferedInputStream));
+    }
+
+    @Override
+    public boolean isDone(final LineIterator lineIterator) {
+        return !lineIterator.hasNext() && activeTopLevelFeatures.isEmpty() && featuresToFlush.isEmpty();
+    }
+
+    @Override
+    public void close(final LineIterator lineIterator) {
+        CloserUtil.close(lineIterator);
+    }
+
+    @Override
+    public TabixFormat getTabixFormat() {
+        return TabixFormat.GFF;
+    }
+
+    /**
+     * Enum for parsing directive lines.  If information in directive line needs to be parsed beyond specifying directive type, decode method should be overriden
+     */
+    enum Gff3Directive {
+
+        VERSION3_DIRECTIVE("##gff-version 3(?:.\\d)*(?:\\.\\d)*$"),
+
+        SEQUENCE_REGION_DIRECTIVE("##sequence-region .+ \\d+ \\d+$") {
+            private int CONTIG_INDEX = 1;
+            private int START_INDEX = 2;
+            private int END_INDEX = 3;
+            @Override
+            public Object decode(final String line) throws IOException {
+                final String[] splitLine = line.split(" +");
+                final String contig = URLDecoder.decode(splitLine[CONTIG_INDEX], "UTF-8");
+                final int start = Integer.valueOf(splitLine[START_INDEX]);
+                final int end = Integer.valueOf(splitLine[END_INDEX]);
+                return new SequenceRegion(contig, start, end);
+            }
+        },
+
+        FLUSH_DIRECTIVE("###$"),
+
+        FASTA_DIRECTIVE("##FASTA$");
+
+        private final String regexPattern;
+
+        Gff3Directive(String regexPattern) {
+            this.regexPattern = regexPattern;
+        }
+
+        public static Gff3Directive toDirective(final String line) {
+            for (final Gff3Directive directive : Gff3Directive.values()) {
+                if(Pattern.matches(directive.regexPattern, line)) {
+                    return directive;
+                }
+            }
+            return null;
+        }
+
+        public Object decode(final String line) throws IOException {
```

Comment:
While that's true currently, I want to leave this as an Object because there are other directives in the spec that I haven't implemented at the moment, which might want to return other object types if implemented in the future.  For example the spec allows for directives indicating particular ontologies, or particular genome builds.